src/PowerPing/ExtensionMethods.cs

```diff
@@ -0,0 +1,43 @@
+﻿/*
+MIT License - PowerPing 
+
+Copyright (c) 2018 Matthew Carney
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace PowerPing
+{
+    static class ExtensionMethods
+    {
+        public static T WaitForResult<T>(this Task<T> task, CancellationToken cancellationToken)
+        {
+            try {
+                task.Wait(cancellationToken);
+            } catch (AggregateException ex) when (ex.InnerExceptions.Count == 1) {
```

Comment:
Task.Wait wraps everything in an AggregateException. I am 'unwrapping' it here so we get the original exception type back. Before my last commit, this was important because the BeginReceive task could throw a SocketException, and there was code at the bottom of SendICMP to catch that SocketException, which wouldn't have worked if it was still wrapped in an AggregateException.